File: yastd/lib/yast2/dbus/service.rb
# frozen_string_literal: true

# Copyright (c) [2022] SUSE LLC
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of version 2 of the GNU General Public License as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, contact SUSE LLC.
#
# To contact SUSE LLC about this file by physical or electronic mail, you may
# find current contact information at www.suse.com.

require "dbus"
require "yast2/dbus/installer"
require "yast2/dbus/language"
require "yast2/dbus/software"
require "yast2/installer"

module Yast2
  module DBus
    # YaST D-Bus service (org.opensuse.YaST)
    #
    # It connects to the system D-Bus and answers requests on `/org/opensuse/YaST/Installer`.
    #
    # @example Running the server
    #   Yast2::DBus::Service.new.run
    #
    # @see Yast2::DBus::Installer
    class Service
      # @return [String] service name
      SERVICE_NAME = "org.opensuse.YaST"

      # @return [String] D-Bus object path
      OBJECT_PATH = "/org/opensuse/YaST/Installer1"

      attr_reader :bus

      def initialize(logger = nil)
        @logger = logger || Logger.new(STDOUT)
        @bus = ::DBus::SystemBus.instance
      end

      # Exports the installer object through the D-Bus service
      def export
        dbus_objects.each { |o| service.export(o) }

        paths = dbus_objects.map(&:path).join(", ")
        logger.info "Exported #{paths} objects"
      end

      def dispatch
        bus.dispatch_message_queue
      end

    private

      attr_reader :logger

      def service
        @service ||= bus.request_service(SERVICE_NAME)
      end

      def dbus_objects
        @dbus_objects ||= [installer_dbus, language_dbus, software_dbus]
      end

      def installer_dbus
        @installer_dbus ||= Yast2::DBus::Installer.new(yast_installer, logger)
      end

      def language_bus
        @language_dbus ||= Yast2::DBus::Language.new(yast_installer, logger)
      end

      def software_bus
        @software_dbus ||= Yast2::DBus::Software.new(yast_installer, logger)
      end

      def installer
        @yast_installer ||= Yast2::Installer.new(logger: logger).tap do |installer|
          # FIXME: do not probe by default
          installer.probe
          installer.dbus_objects = dbus_objects
        end
      end
    end
  end
end

File: rust/agama-utils/src/service.rs
// Copyright (c) [2025] SUSE LLC
//
// All Rights Reserved.
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 2 of the License, or (at your option)
// any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
// more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, contact SUSE LLC.
//
// To contact SUSE LLC about this file by physical or electronic mail, you may
// find current contact information at www.suse.com.

//! Offers a trait to implement an Agama service.
//!
//! An Agama service is composed of, at least, two parts:
//!
//! * The service itself, which holds the configuration and takes care of
//!   performing the changes at installation time. It is private to each
//!   Agama module (agama-l10n, agama-network, etc.). It should implement
//!   the [Service trait].
//! * The handler, which offers an API to talk to the service. It should
//!   implement the [Handler](crate::Handler) trait.

use core::future::Future;
use std::{any, error};
use tokio::sync::mpsc;

/// Implements the basic behavior for an Agama service.
///
/// It is responsible for:
///
/// * Holding the configuration.
/// * Making an installation proposal for one aspect of the system
///   (localization, partitioning, etc.).
/// * Performing the changes a installation time.
/// * Optionally, making changes to the system running Agama
///   (e.g., changing the keyboard layout).
///
/// Usually, a service runs on a separate task and receives the actions to
/// perform through a [mpsc::UnboundedReceiver
/// channel](tokio::sync::mpsc::UnboundedReceiver).
pub trait Service: Send {
    type Err: error::Error;
    type Message: Send;

    /// Returns the service name used for logging and debugging purposes.
    ///
    /// An example might be "agama_l10n::l10n::L10n".
    fn name() -> &'static str {
        any::type_name::<Self>()
    }

    /// Main loop of the service.
    ///
    /// It dispatches one message at a time.
    fn run(&mut self) -> impl Future<Output = ()> + Send {
        async {
            loop {
                let message = self.channel().recv().await;
                let Some(message) = message else {
                    eprintln!("channel closed for {}", Self::name());
                    break;
                };

                if let Err(error) = &mut self.dispatch(message).await {
                    eprintln!("error dispatching command: {error}");
                }
            }
        }
    }

    /// Returns the channel to read the messages from.
    fn channel(&mut self) -> &mut mpsc::UnboundedReceiver<Self::Message>;

    /// Dispatches a message.
    fn dispatch(
        &mut self,
        command: Self::Message,
    ) -> impl Future<Output = Result<(), Self::Err>> + Send;
}
